src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,231 @@
+      int add_component (const char *name);
+      inline int MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx);
+
+    private:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    private:
+      std::string m_module_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_component>> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  class memory_monitor
+  {
+    public:
+      memory_monitor (const char *server_name)
+	: m_aggregater (this), m_server_name (server_name) {}
+      memory_monitor (const memory_monitor &) = delete;
+      memory_monitor (memory_monitor &&) = delete;
+
+      ~memory_monitor () {}
+
+      memory_monitor &operator = (const memory_monitor &) = delete;
+      memory_monitor &operator = (memory_monitor &&) = delete;
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size, bool expand);
+      int sub_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID src, MMON_STAT_ID dest, uint64_t size);
+      int aggregate_module_info (MMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MMON_TRAN_INFO *info, int tran_count);
+      inline int COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+      inline int SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+
+    private:
+      class aggregater
+      {
+	public:
+	  aggregater (memory_monitor *mmon);
+	  aggregater (const aggregater &) = delete;
+	  aggregater (aggregater &&) = delete;
+
+	  ~aggregater () {}
+
+	  aggregater &operator = (const aggregater &) = delete;
+	  aggregater &operator = (aggregater &&) = delete;
+
+	  /* function */
+	  int get_server_info (const MMON_SERVER_INFO &info);
+	  int get_module_info (const MMON_MODULE_INFO &info, int module_index);
+	  int get_transaction_info (const MMON_TRAN_INFO &info, int tran_count);
+
+	private:
+	  /* backlink of memory_monitor class */
+	  memory_monitor *m_mmon;
+      };
+
+    private:
+      std::string m_server_name;
+      std::atomic<uint64_t> m_total_mem_usage;
+      aggregater m_aggregater;
+
+      std::unique_ptr<mmon_module> m_module[MMON_MODULE_LAST];
+  };
+
+  inline int mmon_module::MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+  {
+    return (comp_idx << 16 | subcomp_idx);
+  }
+
+  inline int memory_monitor::COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+  {
+    return (comp_idx_map_val >> 16);
+  }
+
+  inline int memory_monitor::SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+  {
+    return (comp_idx_map_val & MMON_PARSE_MASK);
+  }
```

Comment:
Is there any reason to use upper-case rather than lower-case?

Reply:
Because I think this inline functions are macro-like functions.